### ISSUE TITLE
[FEATURE MODIFICATION] Isometric tilesets support

### ISSUE THREAD

**Issue by Ratysz** [amethyst/amethyst#1971]
# Tasks
- create external crate for tiling logic
- integrate crate into amethyst using ECS and renderer

## Feature

This request concerns the newly added `amethyst_tiles` crate.

## Change

Currently, only orthographic tilesets are supported by `amethyst_tiles`. It should instead be possible to also use it with isometric, or, ideally, arbitrary tilesets.

## Reason

Isometric tilesets are an extremely easy-to-use and, if done not entirely naively, very performant technique for primitive(ish) 3D.

Several established engines already have it ([Unity,](https://docs.unity3d.com/Manual/Tilemap-Isometric-CreateIso.html) [Godot](https://docs.godotengine.org/en/3.0/classes/class_tilemap.html)).

Many games have used it in the past, and many more will likely use it in the future. ([Gnomoria](store.steampowered.com/app/224500/Gnomoria/), [Towns](https://store.steampowered.com/app/221020/Towns/), [Townsmen](https://store.steampowered.com/app/938380/Townsmen__A_Kingdom_Rebuilt/), [Citystate](https://store.steampowered.com/app/774351/Citystate/), [It's Not Cool](https://ratys.itch.io/its-not-cool), ...)

## Impact

Required changes will most likely touch `TileMap` component, it's supporting traits/structs, and relevant shaders.

## Additional Notes

"Arbitrary" tilesets mentioned above refer to projections that are not strictly orthographic or isometric, e.g. trimetric, cavalier, etc.

Essentially non-cubic tilesets - triangular, hexagonal, Penrose, whatever - should also be considered, but such a feature is niche enough to warrant an external crate.

**Comment by jaynus** [comment]
Currently you can achieve this with a perspective projection angled correctly. Is this not sufficient, and instead you want an integrated 2d solution rotating and z-layering the tiles?

**Comment by Ratysz** [comment]
I wasn't able to do that, there are gaps between the tiles, no matter how I rotate the tilemap. Could just be the result of #1970.

**Comment by jaynus** [comment]
Have you tried moving the camera instead?

I'll try to set up a isometric example this weekend, and add any needed features as I do

**Comment by Ratysz** [comment]
I have, now - similar results, plus layering looks off. I don't know enough on the topic to definitively say if it's me doing something wrong, or that's just not how it works, but I imagine neither Godot nor Unity would have dedicated isometric tileset workflows if that was simply a matter of rotating the camera.

Unrelated: tilemap rendering seems to ignore translation component of it's entity's transform. Is that intentional?

**Comment by jaynus** [comment]
@Ratysz I'm not exhibiting any ignoring of translations - you can see a test on #1974 

I'm actively experimenting with some different map projection designs now. I should have something together this weekend.

**Comment by Raymond26** [comment]
Came upon this in a google search - any update or plans @jaynus?

**Comment by Ratysz** [comment]
Ahh, I forgot to bump this. Why does that bot even exist?..

**Comment by CleanCut** [comment]
@Ratysz I can reopen this for you.

**Comment by Benlitz** [comment]
Hello there, I'm having the same problem.
I think the proposed solution, which if I understand properly is to rotate and scale the tilemap, isn't ideal because it deforms the tiles, and it does not allow to use a isometric sprite sheet such the one below. Also it does not allow to do hexagonal tiles (like in a Civilization game)
However, both (iso/hexa) would be easily achievable if the `transform` matrix of `TileMap` was accessible and modifiable. The isometric projection can be achieved with a trivial matrix, and this way tiles can be overlapping, which allows to use the spritesheet below or doing hexagonal tiles in a similar way. 
I'm proposing to either expose the `transform` (to the constructor of `TileMap` maybe?) as `Option`al, or to expose an enum to decide the type of tiling (good cause this can be handled internally, but bad cause limited to what is implemented).

Also, the currently computed transform transform is a bit misplaced in my opinion, a 1x1 tilemap at (0,0) won't be placed at the same location than a single identical sprite placed at (0,0), but maybe I should open another ticket for that.

![tilemap](https://user-images.githubusercontent.com/2251138/101893337-f17e4300-3be7-11eb-95a8-b624fceb1e9a.png)

**Comment by Benlitz** [comment]
Note: I opened a PR to fix the positioning of the map (#2540) and I implemented isometric and hexagonal tiling in another branch (using an `enum` actually) from which I can open another PR once I've cleaned the code and implemented some tests

**Comment by sizeak** [comment]
@Benlitz Are you planning to PR your Isometric support? I would find that quite useful!

**Comment by Benlitz** [comment]
@sizeak yes my implementation works using specs and currently there's an ongoing effort to migrate to legion on the `main` branch (0.16).
Once this is done (looks to be pretty soon), I'll probably migrate my branch as well and do a PR

**Comment by sizeak** [comment]
Excellent, thanks for the quick reply

**Comment by mbrucher** [comment]
Awesome, if we can get in hex support, that would be awesome. Can't wait to see 0.16 as well with legion :)

**Comment by jojolepro** [comment]
Open a pull request whenever ready :)
From now on, we want to be a bit more careful with what we introduce into the engine. As such, if possible, it would be nice if the "core logic" could be in an external crate (which doesn't depend on an ECS) and on the amethyst side the PR would integrate this crate with the ECS and our renderer.
